privacy-protections/request-blocking/main.js

```diff
@@ -0,0 +1,521 @@
+                    } else if (registration.installing) {
+                        registration.installing.addEventListener('statechange', () => {
+                            if (registration.active) {
+                                registration.active.addEventListener('message', () => {
+                                    console.log('client received message');
+                                });
+                                registration.active.postMessage('fetch');
+                            } 
+                        });
+                    }
+                })
+                .catch((error) => {
+                    console.log('Registration failed with ' + error);
+                });
+        }
+    },
+    {
+        category: 'other',
+        id: 'csp-report',
+        description: 'Try sending data via CSP report.',
+        html: `<img src='data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg=='/>`,// causes CSP violation which triggers a report
+        checkAsync: (callback) => {
+            const observer = new PerformanceObserver(observed); 
+
+            const checkResource = resource => {
+                if (resource.name.includes('/csp')) {
+                    if (resource.duration === 0 && resource.nextHopProtocol === '') {
+                        callback('failed');
+                    } else {
+                        callback('loaded');
+                    }
+                    observer.disconnect();
+                }
+            };
+
+            function observed(list) { 
+                list.getEntries().forEach(checkResource);
+            } 
+
+            observer.observe({entryTypes: ["resource", "navigation"]});
+        }
+    }
+];
+
+/**
+ * Test runner
+ */
+function runTests() {
+    startButton.setAttribute('disabled', 'disabled');
+    downloadButton.removeAttribute('disabled');
+    results.results.length = 0;
+
+    tests.forEach(test => {
+        const resultObj = {
+            id: test.id,
+            category: test.category,
+            status: 'not loaded'
+        };
+        results.results.push(resultObj);
+
+        const categoryUl = document.querySelector(`.category-${test.category} ul`);
+
+        const li = document.createElement('li');
+        li.id = `test-${test.category}-${test.id}`;
+        li.innerHTML = `<div class='status' title='not loaded'></div> - ${test.id} - <span class='description'>${test.description || ''}</span>`;
+        const status = li.querySelector('.status');
+
+        if (test.html) {
+            if (typeof test.html === 'string') {
+                const template = document.createElement('template');
+                template.innerHTML = `<hr/>${test.category} - ${test.id} ${test.html}`;
+                debugDiv.appendChild(template.content);
+            } else if (typeof test.html === 'function') {
+                debugDiv.appendChild(test.html());
+            }
```

Comment:
Should we have some kind of default behavior like displaying a console error if a test isn't a string or function? In case in the future something is added and silently isn't tested.

Reply:
Good thinking, added!